dimod/include/dimod/abc.h

```diff
@@ -0,0 +1,1070 @@
+        for (const bias_type& b : biases) {
+            linear_biases_[v] = b;
+            ++v;
+        }
+    }
+
+    /// Set the offset.
+    void set_offset(bias_type offset) { offset_ = offset; }
+
+    /// Set the quadratic bias for the given variables.
+    void set_quadratic(index_type u, index_type v, bias_type bias) {
+        assert(0 <= u && static_cast<size_t>(u) < num_variables());
+        assert(0 <= v && static_cast<size_t>(v) < num_variables());
+
+        enforce_adj();
+
+        if (u == v) {
+            switch (this->vartype(u)) {
+                // unlike add_quadratic, setting is not really defined.
+                case Vartype::BINARY: {
+                    throw std::domain_error(
+                            "Cannot set the quadratic bias of a binary variable with itself");
+                }
+                case Vartype::SPIN: {
+                    throw std::domain_error(
+                            "Cannot set the quadratic bias of a spin variable with itself");
+                }
+                default: {
+                    (*adj_ptr_)[u][v] = bias;
+                    break;
+                }
+            }
+        } else {
+            (*adj_ptr_)[u][v] = bias;
+            (*adj_ptr_)[v][u] = bias;
+        }
+    }
+
+    /// Return the upper bound of variable of `v`.
+    virtual bias_type upper_bound(index_type) const = 0;
+
+    /// Return the vartype of `v`.
+    virtual Vartype vartype(index_type) const = 0;
+
+ protected:
+    explicit QuadraticModelBase(std::vector<bias_type>&& biases)
+            : linear_biases_(biases), adj_ptr_(), offset_(0) {}
+
+    explicit QuadraticModelBase(index_type n) : linear_biases_(n), adj_ptr_(), offset_(0) {}
+
+    /// Increase the size of the model by one. Returns the index of the new variable.
+    index_type add_variable() { return add_variables(1); }
+
+    /// Increase the size of the model by `n`. Returns the index of the first variable added.
+    index_type add_variables(index_type n) {
+        assert(n >= 0);
+        index_type size = num_variables();
+
+        linear_biases_.resize(size + n);
+        if (has_adj()) {
+            adj_ptr_->resize(size + n);
+        }
+
+        return size;
+    }
+
+    /**
+     * Adjust the biases of a single variable to reinterpret it as being a
+     * different vartype.
+     * 
+     * This method is meant to be used by subclasses, since it does not check
+     * or update the actual vartypes of the variables.
+     */
+    void change_vartype(Vartype source, Vartype target, index_type v) {
+        if (source == target) {
+            return;
+        } else if (source == Vartype::SPIN && target == Vartype::BINARY) {
+            if (has_adj()) {
+                for (auto it = (*adj_ptr_)[v].begin(); it != (*adj_ptr_)[v].end(); ++it) {
+                    linear_biases_[it->v] -= it->bias;
+                    (*adj_ptr_)[it->v][v] *= 2;  // log(n)
```

Comment:
Is it adding redundant 0 biases using the [] operator for the neighborhood.

Reply:
That's OK. They are already there.